File: Week Assignment 4 using C/8_pairwise_swap_ll.c
// 8. Pairwise swap nodes of a linked list (swap data for simplicity)
#include <stdio.h>
#include <stdlib.h>

struct Node { int data; struct Node* next; };

struct Node* push(struct Node* head, int data) {
    struct Node* n=(struct Node*)malloc(sizeof(struct Node)); n->data=data; n->next=head; return n;
}

struct Node* pairwiseSwap(struct Node* head) {
    struct Node* cur = head;
    while (cur && cur->next) {
        int tmp = cur->data; cur->data = cur->next->data; cur->next->data = tmp;
        cur = cur->next->next;
    }
    return head;
}

void printList(struct Node* head) {
    printf("[");
    struct Node* t=head; int first=1; while (t){ if (!first) printf(", "); printf("%d", t->data); first=0; t=t->next; } printf("]\n");
}

int main() {
    struct Node* head = NULL; head = push(head,5); head=push(head,4); head=push(head,3); head=push(head,2); head=push(head,1); // 1..5
    printList(head);
    head = pairwiseSwap(head); // expect [2,1,4,3,5]
    printList(head);
    return 0;
}

File: Week Assignement 4 using Python/8_pairwise_swap_ll.py
# 8. Pairwise swap nodes (swap data)
class Node:
    def __init__(self,data):
        self.data=data; self.next=None

def pairwise_swap(head):
    cur = head
    while cur and cur.next:
        cur.data, cur.next.data = cur.next.data, cur.data
        cur = cur.next.next
    return head

def to_list(head):
    res=[]; cur=head
    while cur: res.append(cur.data); cur=cur.next
    return res

if __name__ == "__main__":
    head = Node(1); head.next = Node(2); head.next.next = Node(3); head.next.next.next = Node(4); head.next.next.next.next = Node(5)
    print(to_list(head))
    head = pairwise_swap(head)
    print(to_list(head))
